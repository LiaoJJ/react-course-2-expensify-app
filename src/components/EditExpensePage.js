import React from 'react';
import ExpenseForm from './ExpenseForm';
import {startEditExpense, startRemoveExpense} from '../actions/expenses.js';
import {connect} from 'react-redux';

export class EditExpensePage extends React.Component {
    onSubmit = (expense)=>{
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    };
    onRemove = ()=>{
        this.props.startRemoveExpense({id: this.props.expense.id});
        this.props.history.push('/'); 
    };
    render(){
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>                
                <div className="content-container">
                    <ExpenseForm 
                        className="form"
                        expense={this.props.expense}
                        onSubmit={this.onSubmit}
                    />
                    <button className="button button--remove" onClick={this.onRemove}>Remove Expense</button> 
                </div>

            </div>  
        );
    }
};

const mapStateToProps = (state, props) =>({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id),
});

const mapDispatchToProps = (dispatch) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (data)=>dispatch(startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);