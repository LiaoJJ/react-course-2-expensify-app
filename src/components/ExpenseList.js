import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

// using Provider-connect, there is no parameter passed down from <ExpenseDashboard />

// input props from "mapStateToProps"
// output: render
export const ExpenseList = (props) => (
     <div className="content-container">
         <div className="list-header">
             <div className="show-for-mobile">Expenses</div>
             <div className="show-for-desktop">Expense</div>
             <div className="show-for-desktop">Amount</div>
         </div>
         <div className="list-body">
            {
                props.expenses.length === 0 ? (
                    <div className="list-item list-item--message">
                        <span>No expenses</span>
                    </div>
                ) : (
                    props.expenses.map((expense)=>(
                        <ExpenseListItem key={expense.id} {...expense}/>
                    )
                    )
                )
            }
         </div>
     </div>
 );

 // input: state
 // output: properties that you want to get out of the store
const mapStateToProps = (state)=>{
    return {
        expenses: selectExpenses(state.expenses, state.filter)
    };
}

const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList);


 export default ConnectedExpenseList;