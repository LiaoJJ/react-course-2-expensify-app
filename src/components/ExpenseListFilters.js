import React from 'react';
import {connect} from 'react-redux';
import {DateRangePicker} from 'react-dates';
import {setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate} from '../actions/filters.js';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null,
    };
    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onFocusChange = (calendarFocused) => {
        this.setState(()=>({calendarFocused}));
    };
    onTextChange = (e)=>{
        this.props.setTextFilter(e.target.value);
    }
    onSortChange = (e)=>{
        if(e.target.value==='date'){
            this.props.sortByDate();
        }
        else if(e.target.value==='amount'){
            this.props.sortByAmount();
        }
        
    }
    render(){
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input className="text-input" type="text" value={this.props.filter.text} onChange={this.onTextChange}/>
                    </div>
                    <div className="input-group__item">
                        <select 
                            className="select"
                            value={this.props.filter.sortBy}
                            onChange={this.onSortChange}>
                            <option value="date">Date</option>
                            <option value='amount'>Amount</option>
                        </select> 
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filter.startDate}
                            endDate={this.props.filter.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            showClearDates={true}
                            numberOfMonths={1}
                            isOutsideRange={()=>false}
                        />
                    </div>
                </div>
            </div>
        );
    }
}


const mapStateToProps = (state) => {
    return {
        filter: state.filter
    };
};

const mapDispatchToProps = (dispatch) =>({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    setStartDate: (text) => dispatch(setStartDate(text)),
    setEndDate: (text) => dispatch(setEndDate(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);