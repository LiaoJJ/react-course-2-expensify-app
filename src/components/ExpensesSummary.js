import React from 'react';
import {connect} from 'react-redux';
import {Link} from 'react-router-dom';
import selectExpenses from '../selectors/expenses';
import expense_sum from '../selectors/ExpensesTotal';
import numeral from 'numeral';

// using Provider-connect, there is no parameter passed down from <ExpenseDashboard />

// input props from "mapStateToProps"
// output: render
export const ExpensesSummary = (props) => (
     <div className="page-header">
         <div className="content-container">
            <h1 className="page-header__title">Viewing <span>{props.expenses.length}</span> expense{props.expenses.length>1?'s':''} totaling <span>{numeral(expense_sum(props.expenses)/100).format('$0,0.00')}</span></h1>
            <div className="page-header__actions">
                <Link className="button" to="/create">AddExpense</Link>
            </div>
         </div>
     </div>
 );

 // input: state
 // output: properties that you want to get out of the store
const mapStateToProps = (state)=>{
    return {
        expenses: selectExpenses(state.expenses, state.filter)
    };
}

const ConnectedExpenseList = connect(mapStateToProps)(ExpensesSummary);


 export default ConnectedExpenseList;