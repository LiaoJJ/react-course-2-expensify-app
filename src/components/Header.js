import React from 'react';
import {NavLink} from 'react-router-dom';
import {startLogout} from '../actions/auth';
import { connect } from 'react-redux';

export const Header = ({startLogout}) => (
    <header>
        <h1>Expensify</h1>
        <NavLink activeClassName='is-active' to="/dashboard" exact>Dashboard</NavLink>
        <NavLink activeClassName='is-active' to="/create">Create Expense</NavLink>
        <button onClick={startLogout}>Logout</button>
    </header>
);

const mapDispatchToProps = (dispatch)=>({
    startLogout: ()=> dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);