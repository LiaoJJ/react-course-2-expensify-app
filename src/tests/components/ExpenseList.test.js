import React from 'react';
import {shallow} from 'enzyme';
import {ExpenseList} from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';

test('sada', ()=>{
    const wrapper = shallow(<ExpenseList expenses={expenses}/>);
    expect(wrapper).toMatchSnapshot();
});

test('asdasd', ()=>{
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
});