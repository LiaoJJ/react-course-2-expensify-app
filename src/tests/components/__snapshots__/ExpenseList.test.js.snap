// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asdasd 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ExpenseList
    expenses={Array []}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="list-header"
>
          <div
                    className="show-for-mobile"
          >
                    Expenses
          </div>
          <div
                    className="show-for-desktop"
          >
                    Expense
          </div>
          <div
                    className="show-for-desktop"
          >
                    Amount
          </div>
</div>,
        <div
          className="list-body"
>
          <div
                    className="list-item list-item--message"
          >
                    <span>
                              No expenses
                    </span>
          </div>
</div>,
      ],
      "className": "content-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="show-for-mobile"
>
              Expenses
</div>,
            <div
              className="show-for-desktop"
>
              Expense
</div>,
            <div
              className="show-for-desktop"
>
              Amount
</div>,
          ],
          "className": "list-header",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Expenses",
              "className": "show-for-mobile",
            },
            "ref": null,
            "rendered": "Expenses",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Expense",
              "className": "show-for-desktop",
            },
            "ref": null,
            "rendered": "Expense",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Amount",
              "className": "show-for-desktop",
            },
            "ref": null,
            "rendered": "Amount",
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="list-item list-item--message"
>
            <span>
                        No expenses
            </span>
</div>,
          "className": "list-body",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <span>
              No expenses
</span>,
            "className": "list-item list-item--message",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "No expenses",
            },
            "ref": null,
            "rendered": "No expenses",
            "type": "span",
          },
          "type": "div",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="list-header"
>
            <div
                        className="show-for-mobile"
            >
                        Expenses
            </div>
            <div
                        className="show-for-desktop"
            >
                        Expense
            </div>
            <div
                        className="show-for-desktop"
            >
                        Amount
            </div>
</div>,
          <div
            className="list-body"
>
            <div
                        className="list-item list-item--message"
            >
                        <span>
                                    No expenses
                        </span>
            </div>
</div>,
        ],
        "className": "content-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="show-for-mobile"
>
                Expenses
</div>,
              <div
                className="show-for-desktop"
>
                Expense
</div>,
              <div
                className="show-for-desktop"
>
                Amount
</div>,
            ],
            "className": "list-header",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Expenses",
                "className": "show-for-mobile",
              },
              "ref": null,
              "rendered": "Expenses",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Expense",
                "className": "show-for-desktop",
              },
              "ref": null,
              "rendered": "Expense",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Amount",
                "className": "show-for-desktop",
              },
              "ref": null,
              "rendered": "Amount",
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="list-item list-item--message"
>
              <span>
                            No expenses
              </span>
</div>,
            "className": "list-body",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <span>
                No expenses
</span>,
              "className": "list-item list-item--message",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "No expenses",
              },
              "ref": null,
              "rendered": "No expenses",
              "type": "span",
            },
            "type": "div",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;

exports[`sada 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ExpenseList
    expenses={
        Array [
            Object {
              "amount": 195,
              "createdAt": 0,
              "description": "gum",
              "id": "1",
              "note": "",
            },
            Object {
              "amount": 109500,
              "createdAt": -345600000,
              "description": "Rent",
              "id": "2",
              "note": "",
            },
            Object {
              "amount": 4500,
              "createdAt": 345600000,
              "description": "Credit Card",
              "id": "3",
              "note": "",
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="list-header"
>
          <div
                    className="show-for-mobile"
          >
                    Expenses
          </div>
          <div
                    className="show-for-desktop"
          >
                    Expense
          </div>
          <div
                    className="show-for-desktop"
          >
                    Amount
          </div>
</div>,
        <div
          className="list-body"
>
          <ExpenseListItem
                    amount={195}
                    createdAt={0}
                    description="gum"
                    id="1"
                    note=""
          />
          <ExpenseListItem
                    amount={109500}
                    createdAt={-345600000}
                    description="Rent"
                    id="2"
                    note=""
          />
          <ExpenseListItem
                    amount={4500}
                    createdAt={345600000}
                    description="Credit Card"
                    id="3"
                    note=""
          />
</div>,
      ],
      "className": "content-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="show-for-mobile"
>
              Expenses
</div>,
            <div
              className="show-for-desktop"
>
              Expense
</div>,
            <div
              className="show-for-desktop"
>
              Amount
</div>,
          ],
          "className": "list-header",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Expenses",
              "className": "show-for-mobile",
            },
            "ref": null,
            "rendered": "Expenses",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Expense",
              "className": "show-for-desktop",
            },
            "ref": null,
            "rendered": "Expense",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Amount",
              "className": "show-for-desktop",
            },
            "ref": null,
            "rendered": "Amount",
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <ExpenseListItem
              amount={195}
              createdAt={0}
              description="gum"
              id="1"
              note=""
/>,
            <ExpenseListItem
              amount={109500}
              createdAt={-345600000}
              description="Rent"
              id="2"
              note=""
/>,
            <ExpenseListItem
              amount={4500}
              createdAt={345600000}
              description="Credit Card"
              id="3"
              note=""
/>,
          ],
          "className": "list-body",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "function",
            "props": Object {
              "amount": 195,
              "createdAt": 0,
              "description": "gum",
              "id": "1",
              "note": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "function",
            "props": Object {
              "amount": 109500,
              "createdAt": -345600000,
              "description": "Rent",
              "id": "2",
              "note": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "3",
            "nodeType": "function",
            "props": Object {
              "amount": 4500,
              "createdAt": 345600000,
              "description": "Credit Card",
              "id": "3",
              "note": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="list-header"
>
            <div
                        className="show-for-mobile"
            >
                        Expenses
            </div>
            <div
                        className="show-for-desktop"
            >
                        Expense
            </div>
            <div
                        className="show-for-desktop"
            >
                        Amount
            </div>
</div>,
          <div
            className="list-body"
>
            <ExpenseListItem
                        amount={195}
                        createdAt={0}
                        description="gum"
                        id="1"
                        note=""
            />
            <ExpenseListItem
                        amount={109500}
                        createdAt={-345600000}
                        description="Rent"
                        id="2"
                        note=""
            />
            <ExpenseListItem
                        amount={4500}
                        createdAt={345600000}
                        description="Credit Card"
                        id="3"
                        note=""
            />
</div>,
        ],
        "className": "content-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="show-for-mobile"
>
                Expenses
</div>,
              <div
                className="show-for-desktop"
>
                Expense
</div>,
              <div
                className="show-for-desktop"
>
                Amount
</div>,
            ],
            "className": "list-header",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Expenses",
                "className": "show-for-mobile",
              },
              "ref": null,
              "rendered": "Expenses",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Expense",
                "className": "show-for-desktop",
              },
              "ref": null,
              "rendered": "Expense",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Amount",
                "className": "show-for-desktop",
              },
              "ref": null,
              "rendered": "Amount",
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <ExpenseListItem
                amount={195}
                createdAt={0}
                description="gum"
                id="1"
                note=""
/>,
              <ExpenseListItem
                amount={109500}
                createdAt={-345600000}
                description="Rent"
                id="2"
                note=""
/>,
              <ExpenseListItem
                amount={4500}
                createdAt={345600000}
                description="Credit Card"
                id="3"
                note=""
/>,
            ],
            "className": "list-body",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "function",
              "props": Object {
                "amount": 195,
                "createdAt": 0,
                "description": "gum",
                "id": "1",
                "note": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "2",
              "nodeType": "function",
              "props": Object {
                "amount": 109500,
                "createdAt": -345600000,
                "description": "Rent",
                "id": "2",
                "note": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "3",
              "nodeType": "function",
              "props": Object {
                "amount": 4500,
                "createdAt": 345600000,
                "description": "Credit Card",
                "id": "3",
                "note": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;
